mmcv/ops/csrc/common/cuda/softmax_focal_loss_cuda_kernel.cuh

```diff
@@ -10,62 +10,125 @@
 
 template <typename T>
 __global__ void softmax_focal_loss_forward_cuda_kernel(
-    const int nthreads, const T* softmax, const int64_t* target,
-    const T* weight, T* output, const T gamma, const T alpha,
-    const int num_classes) {
+    const int nthreads, const T* input, const int64_t* target,
+    const T* weight, T* output, T* log_softmax_prob,
+    const T gamma, const T alpha, const int num_classes) {
   CUDA_1D_KERNEL_LOOP(index, nthreads) {
-    int64_t label = target[index];
-    T pred = softmax[index * num_classes + label];
+    const int n = index / num_classes;
+    const int c = index % num_classes;
+
+    // log_softmax for numerical stability
+    const int start = n * num_classes;
+    const int end = start + num_classes;
+
+    const T max_val_along_class = [&] {
+      T max_val_along_class = -FLT_MAX;
+      for(int c_idx = start; c_idx < end; ++c_idx) {
+        max_val_along_class = max(max_val_along_class, input[c_idx]);
```

Comment:
As input will be loaded multiple times, better cache it in shared memory or load it with ldg.

Reply:
And max_val_along_class can be shared too.

Reply:
Tried to add `__ldg` to them in commit [7081ba7](https://github.com/open-mmlab/mmcv/pull/2893/commits/7081ba742552da750a91990310c70bbaf378e515), but got `CUDA error: operation not supported on global/shared address space` while `pytest`. This error looks strange, and my guess is that the L1/shared memory size is 48kB and it is not enough during runtime.

If we use `__shared__` memory, there should be a `__syncthreads()`. The performance advantage of this `log_softmax` implementation may decrease. Thus, maybe it is better to calculate `log_softmax` by PyTorch in Python part, and then pass it to the cuda kernel function. Any comment about this?

Reply:
> Thus, maybe it is better to calculate log_softmax by PyTorch in Python part, and then pass it to the cuda kernel function
Agree